### PR TITLE
Fix opt-cross-module using non version safe tokens

### PR DESCRIPTION
**Pull request by davidwrighton**
Fix issue where in some cases the compiler would generate references using tokens from modules not in the current module set.

Also add an assert to detect cases where this might happen

### CODE REVIEW

File: src/coreclr/tools/aot/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRun/ModuleTokenResolver.cs [hunk 1]
```diff
@@ -246,6 +247,13 @@ private void AddModuleTokenForFieldReference(TypeDesc owningType, ModuleToken to
         // be consistent in all runs of the compiler
         void SetModuleTokenForTypeSystemEntity<T>(ConcurrentDictionary<T, ModuleToken> dictionary, T tse, ModuleToken token)
         {
+            // We can only use tokens from the manifest mutable module or from one of the modules that versions
+            // with the current compilation. Any other module tokens may change while the code executes
+            if (token.Module != _manifestMutableModule && _compilationModuleGroup.VersionsWithModule((ModuleDesc)token.Module));
```
Comment (teo-tsirpanis) [inline src/coreclr/tools/aot/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRun/ModuleTokenResolver.cs:252]:
```suggestion
            if (token.Module != _manifestMutableModule && _compilationModuleGroup.VersionsWithModule((ModuleDesc)token.Module))
```

File: src/coreclr/tools/aot/ILCompiler.ReadyToRun/JitInterface/CorInfoImpl.ReadyToRun.cs [hunk 2]
```diff
@@ -778,7 +776,27 @@ public void CompileMethod(MethodWithGCInfo methodCodeNodeNeedingCode, Logger log
...
                     var resolver = _compilation.NodeFactory.Resolver;
-                    resolver.AddModuleTokenForMethod(MethodBeingCompiled, new ModuleToken(ecmaMethod.Module, ecmaMethod.Handle));
+                    if (_compilation.CompilationModuleGroup.VersionsWithModule(ecmaMethod.Module))
+                    {
+                        resolver.AddModuleTokenForMethod(MethodBeingCompiled, new ModuleToken(ecmaMethod.Module, ecmaMethod.Handle));
+                    }
+                    else
+                    {
+                        // Cross module optimization scenario. If we aren't able to come up with a token for the method being compiled, then force it to be
+                        // computed.
```
Comment (mangod9) [inline src/coreclr/tools/aot/ILCompiler.ReadyToRun/JitInterface/CorInfoImpl.ReadyToRun.cs:786]: assume the computation is deterministic based on the methodName, returnValue etc?

Comment (davidwrighton) [reply src/coreclr/tools/aot/ILCompiler.ReadyToRun/JitInterface/CorInfoImpl.ReadyToRun.cs:786]: Yes. Computation is deterministic as setting the `_ilBodiesNeeded` variable will defer work until single threaded, and then will sort the list and fill stuff in deterministically.

File: src/coreclr/tools/aot/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRun/ModuleTokenResolver.cs (hunk 1 above)
Comment (davidwrighton) [reply src/coreclr/tools/aot/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRun/ModuleTokenResolver.cs:252]: Sigh. This is what I get for refactoring just before committing.

File: src/coreclr/tools/aot/ILCompiler.ReadyToRun/JitInterface/CorInfoImpl.ReadyToRun.cs (hunk 2 above)
Comment (mangod9) [reply src/coreclr/tools/aot/ILCompiler.ReadyToRun/JitInterface/CorInfoImpl.ReadyToRun.cs:786]: sounds good.

**Review by mangod9 — APPROVED** [review]

**Pull request merged**